docs/design/ha_master.md

```diff
@@ -0,0 +1,255 @@
+  (as explained [here](https://coreos.com/etcd/docs/latest/getting-started-with-etcd.html)).
+* We will introduce a DNS name for a cluster and optionally add a load balancer
+  in front of apiservers to spread the load among all the replicas.
+* Controller manager, scheduler & cluster autoscaler will use lease mechanism and
+  only a single instance will be an active master. All other will be waiting in a standby mode.
+* All add-on managers will work independently and each of them will try to keep add-ons in sync
+
+# Detailed design
+
+## VM naming in GCE
+
+VMs with master replicas will have the following names:
+* first VM - `${MASTER_NAME}`
+* next VMs - `${MASTER_NAME}-<hash>`, where `<hash>` is a 3-characters long alphanumeric string.
+
+It will be ok to delete VM without the suffix.
+
+If user will run deployment scripts with a different `MASTER_NAME` variable than earlier,
+it will fail with an error that there are no masters to replicate.
+
+## Components
+
+### etcd
+
+```
+Note: This design for etcd clustering is quite pet-set like - each etcd
+replica has its name which is explicitly used in etcd configuration etc. In
+medium-term future we would like to have the ability to run masters as part of
+autoscaling-group (AWS) or managed-instance-group (GCE) and add/remove replicas
+automatically. This is pretty tricky and this design does not cover this.
+It will be covered in a separate doc.
+```
+
+```
+Note: This section assumes we are using etcd v2.2; it will have to be revisited during upgrade
+to v3, e.g. different command line options.
+```
+
+All etcd instances will be clustered together and one of them will be an elected master.
+In order to commit any change quorum of the cluster will have to confirm it. Etcd will be
+configured in such a way that all writes and reads will go through the master (requests
+will be forwarded by the local etcd server such that it’s invisible for the user). It will
+affect latency for all operations, but it should not increase by much more than the network
+latency between master replicas (latency between GCE zones with a region is < 10ms).
+
+Currently etcd exposes port only using localhost interface. In order to allow clustering
+and inter-VM communication we will also have to use public interface. To secure the
+communication we will configure firewall rules to only allow traffic on etcd port from
+other master machines.
+
+Another option to secure etcd cluster communication would be to use SSL (as described
+[here](https://coreos.com/etcd/docs/latest/security.html)), but it would increase CPU usage
+on master machines, which is already a bottleneck in large clusters. In many
+use cases it is not required so we will not offer this in the first version.
+
+When generating command line for etcd we will always assume it’s part of a cluster
+(initially of size 1) and list all VMs for kubernetes masters (see VM naming section).
```

Comment:
We will list all existing VMs/master replicas. This is needed to properly configure initial list of etcd cluster members.

Pragraph rephrased.